### ISSUE TITLE
Better allocator for particle property data

### ISSUE THREAD

**Issue by bangerth** [geodynamics/aspect#1215]
The only dynamically allocated memory in the `Particle` class is the `properties` object, which is a `std::vector<double>`. This has always bothered me, because we can have so many particles in a simulation and because each of them allocates a chunk of exactly the same length. There must be simpler ways to do this that (i) don't quite fragment memory as much, and (ii) allow for grouping this data closer together in memory for better cache efficiency.

My original idea was to give each particle object a reference to a central pool upon construction where it can get a handle to a block of size `n_properties` doubles. Instead of a pointer, it would then store a handle to it (i.e., typically an integer index; the pool would then translate this into a concrete pointer). This has a number of advantages:
* The pool would just allocate one large chunk of memory (e.g., a `std::vector<double>`) and the index would be a starting index into this vector. Because we don't store pointers, the pool could re-size the vector if needed, or keep a collection of vectors if it wants to.
* It would not be terrible difficult to reorder the elements in the pool for better cache efficiency after every time particles move between cells, if that's what we want. It would just require one additional permutation array in the pool (in essence, a lookup table that translates from handle to starting index). The reordering could also compress the array to allow for particles that move between cells.
On would have to come up with an implementation of such a pool, but because it only stores fixed size objects, I don't think that would be overly complicated.

A class interface for such a pool class could look like this:
```
  class PropertiesPool {
    public:
    struct Handle {
       private:
          std::size_t index;
    };

  PropertiesPool (const unsigned int n_properties_per_slot);

  Handle allocate_properties_array ();   // return a handle to one slot
  void deallocate_properties_array (Handler &);

  dealii::ArrayView<double> get_properties (const Handler &) const;  // translate to actual address

  private:
    std::vector<double> memory_pool;    // size == N_slots * n_properties_per_slot;

    // possibly a translation table from index to memory pool location
    std::vector<std::size_t> translation_table; // size == N_slots
  };
```
The `Particles::World` class would then have one of these pools, and would pass it to `Particle` objects upon construction. The `Particle` objects store a reference to the pool to do the translation when necessary, and a `Handle` for the memory they allocated. (Or, alternatively, would not store the reference to the pool if they don't ever need to do the translation, and `get_properties()` would just return the `Handle`).

An alternative would be to use the `allocator` template argument of the `std::vector<double>` class with which we currently store the properties in each particle. We would then have to write our own allocator, which would in essence again work as a pool like above by keeping a reference to a pool and drawing from it. Writing such an allocator class isn't terribly difficult -- it just has to have the same interface as described here: http://en.cppreference.com/w/cpp/memory/allocator . Allocators can describe pointer types that would in essence work like the `Handle` class above.

I think this may be the conceptually more elegant solution, primarily because it hooks into existing structures with support in the language, but it may not be easier to implement or use. I don't know which approach to suggest -- opinions would be welcome.

**Comment by gassmoeller** [comment]
I have tried to profile the cache usage and the effect of the dynamic memory allocation on a typical model setup. The test model has 2500 dofs and 10,000 particles in a 2D box geometry and tries to touch every aspect of the algorithm (including remeshing and parallel transfer between 2 processes). 

It seems the dynamic allocation of particle properties takes about 50% of the time spent in the Particle constructor (and the same time when it is resized and initialized), but since the overall time spent in the constructor is very small compared to all other parts of the algorithm (<1% of the particle runtime) I do not expect a major effect of a more efficient implementation. The same holds true for the destructor. Also in turns of L1 and LL cache misses the particle code seems to behave quite well, generating much less relative cache misses than its relative time. That being said I have no experience profiling cache usage, and I certainly agree that using a better implementation would speed up particle generation and destruction (and since they are often constructed and destructed when they move to different cells and domains that might be significant for some applications). Based on the profiling I did so far, I am not expecting much of a performance gain compared to the current implementation for ASPECT though.

Concerning the approach: I agree that using language standards instead of homegrown solutions usually is the way to go. However, I currently have two problems with the allocator approach:
1. I do not quite see how this would need to look like in practice. Would `World` own one object of this allocator type, and the Particles use this object to allocate memory for their own `std::vector<double>`? The allocator would then 'own' the memory of all particles?
2. I think our allocator would require a state (e.g. which memory places are used, which are free, where are holes), but as far as I can see, state carrying allocators are only allowed since C++11. Is there a way around this, or does this disqualify the allocator approach for us?

For the PropertyPool solution I see how it should look like, and could implement a draft that we could discuss, if you agree to proceed with that.

**Comment by bangerth** [comment]
Yes, the question of state of the allocator is something that also stumped me. The allocator needs to keep a reference to the pool. If that can only be done with C++11, then that's a poor solution.

The allocator idea has the advantage that the changes to the codes downstream that need to get a reference to the vector would be relatively limited and only consists of adding a second template argument to a bunch of places. The other idea would require switching wholesale from the `std::vector` to an `ArrayView`.

If you're interested, yes, go ahead with a mockup or draft. It's relatively low on my list of priorities for ASPECT, but it's also something that will become more difficult to do the longer the existing code is already out there.

**Comment by bangerth** [comment]
Now in #1301.

**Comment by bangerth** [comment]
We may want to postpone this until after 2.0 when we can rely on C++11 and therefore have stateful allocators.
